src/frontend/src/binder/expr/function.rs

```diff
@@ -158,13 +230,82 @@ impl Binder {
             )
         {
             use crate::catalog::function_catalog::FunctionKind::*;
-            match &func.kind {
-                Scalar { .. } => return Ok(UserDefinedFunction::new(func.clone(), inputs).into()),
-                Table { .. } => {
-                    self.ensure_table_function_allowed()?;
-                    return Ok(TableFunction::new_user_defined(func.clone(), inputs).into());
+            if func.language == "sql" {
+                // This represents the current user defined function is `language sql`
+                let parse_result =
+                    risingwave_sqlparser::parser::Parser::parse_sql(func.identifier.as_str());
+                if let Err(ParserError::ParserError(err)) | Err(ParserError::TokenizerError(err)) =
+                    parse_result
+                {
+                    // Here we just return the original parse error message
+                    return Err(ErrorCode::InvalidInputSyntax(err).into());
+                }
+                debug_assert!(parse_result.is_ok());
+
+                // We can safely unwrap here
+                let ast = parse_result.unwrap();
+
+                let mut clean_flag = true;
+
+                // We need to check if the `udf_context` is empty first, consider the following example:
+                // - create function add(INT, INT) returns int language sql as 'select $1 + $2';
+                // - create function add_wrapper(INT, INT) returns int language sql as 'select add($1, $2)';
+                // - select add_wrapper(1, 1);
+                // When binding `add($1, $2)` in `add_wrapper`, the input args are [$1, $2] instead of
+                // the original [1, 1], thus we need to check `udf_context` to see if the input
+                // args already exist in the context. If so, we do NOT need to create the context again.
+                // Otherwise the current `udf_context` will be corrupted.
+                if self.udf_context.is_empty() {
+                    // The actual inline logic for sql udf
+                    if let Ok(context) = create_udf_context(&args, &Arc::clone(func)) {
+                        self.udf_context = context;
+                    } else {
+                        return Err(ErrorCode::InvalidInputSyntax(
+                            "failed to create the `udf_context`, please recheck your function definition and syntax".to_string()
+                        )
+                        .into());
+                    }
+                } else {
+                    // If the `udf_context` is not empty, this means the current binding
+                    // function is not the root binding sql udf, thus we should NOT
+                    // clean the context after binding.
+                    clean_flag = false;
```

Comment:
I'm feeling that it's time to create a new binder to do recursive scoped binding, which could simplify state management. 🤔

Reply:
Maybe add an extra field in binder, e.g., `udf_scope_binder: Option<Binder>`?
But I'm worrying the extra cost of adding a new binder when using sql udf, it should be acceptable though 🤣

Reply:
No, I mean creating a new binder on the stack (with only function parameters as available identifiers) and calling the `bind` method. 
The cost can be ignored in the binding stage I think. 😄

Reply:
Will put this to later PRs. :clown_face: